akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -854,15 +854,21 @@ private[stream] class GroupedWithin[T](n: Int, d: FiniteDuration) extends GraphS
   }
 }
 
-private[stream] class Delay[T](d: FiniteDuration, strategy: DelayOverflowStrategy) extends SimpleLinearGraphStage[T] {
-
+private[stream] final class Delay[T](d: FiniteDuration, strategy: DelayOverflowStrategy) extends SimpleLinearGraphStage[T] {
+  override def initialAttributes: Attributes = DefaultAttributes.delay
+  private[this] val timerName = "DelayedTimer"
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
-    val size = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+    val size =
+      inheritedAttributes.getAttribute(classOf[InputBuffer]) match {
+        case None                        ⇒ throw new IllegalStateException(s"Couldn't find InputBuffer Attribute for $this")
+        case Some(InputBuffer(min, max)) ⇒ max
+      }
+
     val buffer = FixedSizeBuffer[(Long, T)](size) // buffer has pairs timestamp with upstream element
-    val timerName = "DelayedTimer"
     var willStop = false
 
     setHandler(in, handler = new InHandler {
+      //FIXME rewrite into distinct strategies f
```

Comment:
I'll clean up the comment, I mean that we shouldn't go through that if-table for every element when we're full. That `DelayOverflowStrategy` won't change.